src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraph.java

```diff
@@ -0,0 +1,464 @@
+        if (nodeKeys.size() != nodesCollection.size()) {
+            final Set<String> badNodes = nodesCollection.stream()
+                    .map(CacheNode::getKey)
+                    .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()))
+                    .entrySet().stream()
+                    .filter(entry -> entry.getValue() > 1)
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toSet());
+
+            throw new IllegalArgumentException("The node list contains nodes with repeated keys: " +
+                badNodes.stream().collect(Collectors.joining(", ")));
+        }
+        nodesMap = nodesCollection.stream().collect(Collectors.toMap(CacheNode::getKey, Function.identity()));
+
+        /* create maps for descendents, parents, and tags */
+        final Map<String, Set<String>> immediateDescendents = new HashMap<>();
+        final Map<String, Set<String>> immediateParents = new HashMap<>();
+        final Map<String, Set<String>> initialTags = new HashMap<>();
+        nodeKeys.forEach(key -> {
+            immediateDescendents.put(key, new HashSet<>());
+            immediateParents.put(key, new HashSet<>());
+            initialTags.put(key, new HashSet<>());
+        });
+
+        /* immediate parents, descendents and tags */
+        nodeKeys.stream().forEach(node -> {
+            nodesMap.get(node).getParents().stream().forEach(parent -> immediateDescendents.get(parent).add(node));
+            immediateParents.get(node).addAll(nodesMap.get(node).getParents());
+            initialTags.get(node).addAll(nodesMap.get(node).getTags());
+        });
+
+        /* build the graph */
+        cgs = new ComputableGraphStructure(nodeKeys, initialTags, immediateParents, immediateDescendents);
+    }
+
+    /**
+     * A private constructor (used by duplicators)
+     * @param nodesMap a previously constructed key -> {@link CacheNode} map
+     * @param cgs a previously constructed {@link ComputableGraphStructure}
+     */
+    private ImmutableComputableGraph(@Nonnull final Map<String, CacheNode> nodesMap,
+                                     @Nonnull final ComputableGraphStructure cgs,
+                                     final boolean cacheAutoUpdate) {
+        this.nodesMap = nodesMap;
+        this.cgs = cgs;
+        this.cacheAutoUpdate = cacheAutoUpdate;
+    }
+
+    /**
+     * Sets the value of a primitive cache node
+     * @param nodeKey the key of the node
+     * @param newValue the new value of the node (note: it will not be duplicated; the user is responsible for duplicating)
+     * @return a new instace of {@link ImmutableComputableGraph} with refernce to unchanged nodes and duplicated
+     *         changed nodes
+     * @throws IllegalArgumentException if the node does not exist
+     * @throws UnsupportedOperationException if the node is non-primitive
+     */
+    public ImmutableComputableGraph setValue(@Nonnull final String nodeKey,
```

Comment:
Yes, ICG is meant to be immutable. In theory, the RDD elements are assumed to be made of immutable objects, even though Java allows mutation (... thanks Obama!!). Mutating older nodes in the Spark DAG will result in indeterministic behavior and impossible-to-catch errors if and when a container is lost and some of the RDD entries need to be recomputed from older ones and the DAG.

The `set(...)` method instantiates a new ICG, populates unchanged cache nodes by reference, outdates affected nodes, and creates a `CacheNode` instance for updated note.